A41-xds-rbac.md

```diff
@@ -62,107 +62,134 @@ Supporting the RBAC HTTP filter on server-side leverages [A36: xDS-Enabled
+The `Content-Type` provided by the client must be used; not a hard-coded value.
 Binary headers are represented in their base64-encoded form, although we rarely
-expect binary header matchers.
+expect binary header matchers other than presence-checking.
 
-TODO: Clearly define differences compared to client-side. Align with client-side
-routing (maybe changing behavior) to allow reusing code, have a strong grip on
-client vs server differences, and notice 'gotcha's.  Content-Type is probably
-not a problem on server-side.
+[hop-by-hop headers]: https://datatracker.ietf.org/doc/html/rfc7230#section-6.1
+[rfc7540 connection header]: https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2
 
 As documented for `HeaderMatcher`, Envoy aliases `:authority` and `Host` in its
 header map implementation, so they should be treated equivalent for the RBAC
 matchers; there must be no behavior change depending on which of the two header
 names is used in the _RBAC policy_.
 
-TODO: We've had no reason to care about :scheme in our APIs and so is not
-present in all languages. It could be inferred from whether the connection used
-security, but that's technically different from :scheme, as if you go through a
-TLS-terminating proxy and use plaintext to the backend the :scheme is used to
-re-create the original URL which would be 'https', not 'http'. This may be a
-problem for Java and Go. We can maybe copy whatever Envoy does for HTTP/1
-traffic. You might think :scheme doesn't matter, and maybe it doesn't, but it
-might be used to detect "is the client using TLS". Although note that in that
-case such a check is trusting the client to provide the correct value.
-
-TODO: Need to audit that server implementations check that :method=POST before
-accepting the RPC. This was already a requirement but it is an appropriate time
-to double-check.
-
-TODO: Determine how to handle _requests_ (not policies) that have Host header,
-since gRPC generally doesn't observe this header yet the RBAC policy could.
-Options: deny such requests, move Host header to :authority, drop Host header.
-Host and :authority could also disagree, and that needs to be handled.
+The core gRPC implementation (not just xDS or RBAC) must observe both :authority
+and Host headers. If :authority is missing, Host must be renamed to :authority.
+If :authority is present, Host must be discarded. If multiple Host headers or
+multiple :authority headers are present, the request must be rejected with an
+HTTP status code 400 as required by Host validation in RFC 7230 §5.4, gRPC
+status code INTERNAL, or RST_STREAM with HTTP/2 error code PROTOCOL_ERROR. These
+restrictions and behavior produce a singular, unambiguous authority for every
+request to be used by RBAC and the application itself.
 
-TODO: Should MetadataMatcher's NullMatch match? Need to check not present vs
-present-but-null behavior.
+`metadata` will never match as `ValueMatcher` can only match if the value is
```

Comment:
I think the discussion of `ValueMatcher` and `NullMatch` is irrelevant and a little confusing.  It would probably be better to just say something like:

`metadata` will never match, because it refers to Envoy metadata, which is not the same thing as gRPC metadata.  Envoy metadata is generic state shared between filters, which has no gRPC equivalent.

Reply:
> metadata will never match, because it refers to Envoy metadata, which is not the same thing as gRPC metadata.

The problem is that's not true. For header matching, for example, you can use `not` so you could check that something _isn't_ present which _could_ be true for gRPC. The only thing that makes this safe is there is no way to match a non-existent value, which is what this sentence is trying to convey.

Reply:
Similar to my comment about the `te` header above, I'm assuming the intention here is to always treat metadata as non-existing, but I'm still finding the wording here hard to follow.  I suggest something like the following:

"""
Note that in RBAC, `metadata` refers to Envoy metadata, which is not the same thing as gRPC metadata.  Envoy metadata is generic state shared between filters, which has no gRPC equivalent.  Therefore, RBAC implementations in gRPC will treat all `metadata` as non-existing.
"""